File: grind-169/20.Valid-Parentheses/valid-parentheses.cpp
/*
 * unordered_map: key, value cannot be switched
 */

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        unordered_map<char, char> bracketMap{
            {')', '('},			// cannot be switched
            {']', '['},
            {'}', '{'},
        };
        
        for (char i : s) {
            if (i == ')' || i == ']' || i == '}') {
                if (st.empty() || st.top() != bracketMap[i])
                    return false;
                st.pop();
            } else if (i == '(' || i == '[' || i == '{') {
                st.push(i);
            }
        }
        return st.empty();
    }
};

File: grind-169/20.Valid-Parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        bracket = {')':'(', ']':'[', '}':'{'}
        for c in s:
            if c in bracket:
                top = stack.pop() if stack else '0'
                if top != bracket[c]:
                    return False
            else:
                stack.append(c)
        return not stack
